rectools/models/candidate_ranking.py

```diff
@@ -0,0 +1,637 @@
+# flake8: noqa
+# TODO: docstrings
+
+import typing as tp
+from functools import reduce
+
+import attr
+import numpy as np
+import pandas as pd
+from catboost import CatBoostClassifier, CatBoostRanker, Pool
+
+from rectools import Columns
+from rectools.dataset import Dataset
+from rectools.dataset.identifiers import ExternalIds
+from rectools.model_selection import Splitter
+from rectools.models.base import ErrorBehaviour, ModelBase, NotFittedError
+
+
+@tp.runtime_checkable
+class ClassifierBase(tp.Protocol):
+    def fit(self, *args: tp.Any, **kwargs: tp.Any) -> None: ...
+
+    def predict_proba(self, *args: tp.Any, **kwargs: tp.Any) -> np.ndarray: ...
+
+
+@tp.runtime_checkable
+class RankerBase(tp.Protocol):
+    def fit(self, *args: tp.Any, **kwargs: tp.Any) -> None: ...
+
+    def predict(self, *args: tp.Any, **kwargs: tp.Any) -> np.ndarray: ...
+
+
+class Reranker:
+    def __init__(
+        self,
+        model: tp.Union[ClassifierBase, RankerBase],
+        fit_kwargs: tp.Optional[tp.Dict[str, tp.Any]] = None,
+    ):
+        self.model = model
+        self.fit_kwargs = fit_kwargs
+
+    def prepare_fit_kwargs(self, candidates_with_target: pd.DataFrame) -> tp.Dict[str, tp.Any]:
+        candidates_with_target = candidates_with_target.drop(columns=Columns.UserItem)
+
+        fit_kwargs = {
+            "X": candidates_with_target.drop(columns=Columns.Target),
+            "y": candidates_with_target[Columns.Target],
+        }
+
+        if self.fit_kwargs is not None:
+            fit_kwargs.update(self.fit_kwargs)
+
+        return fit_kwargs
+
+    def fit(self, candidates_with_target: pd.DataFrame) -> None:
+        fit_kwargs = self.prepare_fit_kwargs(candidates_with_target)
+        self.model.fit(**fit_kwargs)
+
+    def rerank(self, candidates: pd.DataFrame) -> pd.DataFrame:
+        reco = candidates[Columns.UserItem].copy()
+        x_full = candidates.drop(columns=Columns.UserItem)
+
+        # TODO: think about it
+        # x_full = x_full[self.model.feature_names_]
+
+        if isinstance(self.model, ClassifierBase):
+            reco[Columns.Score] = self.model.predict_proba(x_full)[:, 1]
+        elif isinstance(self.model, RankerBase):
+            reco[Columns.Score] = self.model.predict(x_full)
+        else:
+            raise ValueError("Got unexpected model_type")
+        reco.sort_values(by=[Columns.User, Columns.Score], ascending=False, inplace=True)
+        return reco
+
+
+class CatBoostReranker(Reranker):
+    def __init__(
+        self,
+        model: tp.Union[CatBoostClassifier, CatBoostRanker] = CatBoostRanker(verbose=False),
```

Comment:
It's weird linters are silent here.  
But setting a mutable object as a default value is a really bad idea

Reply:
Fixed